### ISSUE TITLE
Decided on how to make IPs of containers available in LAN

### ISSUE THREAD

**Issue by johannesduesing** [delphi-hub/delphi-registry#24]
There are two ways of accessing a container **form anywhere inside the LAN** that is deployed on a certain host:
 1. Map the relevant port of the container to a port of the docker-host, than from inside your LAN you have to call *docker-host*:*mapped-port*/*command*. This  would be not that clean, since the host of all instances would be the docker-host.
2. Bridge the docker-network to the LAN, so every container is reachable via its container-IP. This seems to be possible, following this explanation: https://blog.jessfraz.com/post/ips-for-all-the-things/

**Comment by johannesduesing** [comment]
So-called [Macvlan networks](https://docs.docker.com/network/macvlan/) seem to be a good solution for this problem. They allow containers to receive their own MAC addresses and make them part of the LAN by briding to the host interface. A useful introduction can be found [here](https://docs.docker.com/network/network-tutorial-macvlan/#bridge-example). 
I was able to make a webapp container reachable from another machine on the LAN by following [this](https://hicu.be/docker-networking-macvlan-bridge-mode-configuration) tutorial. There is one major drawback: Containers that use a macvlan bridge can not connect to their host interface. This means that, if the registry and docker host are on the same machine, the containers will not be able to connect to the registry.
Maybe there is a way around that problem ( as described [here](http://blog.oddbit.com/2018/03/12/using-docker-macvlan-networks/) ), if not we have to discuss whether this is feasible or not.

**Comment by 24santoshr** [comment]
I saw a [post](https://stackoverflow.com/questions/46951569/how-to-connect-containers-on-two-different-hosts-connected-by-lan) mentioning the approaches we could use for solving this issue. 
I will go through these approaches and will do a poc. 
Hopefully, one of these approaches should work.

**Comment by johannesduesing** [comment]
After we tried to implement the macvlan solution for quite some time and did not get any consistent results, i think it is reasonable to go for a reverse-proxy to make containers reachable from outside their docker host.
We settled for [Traefik](https://traefik.io/), which was quite easy to setup following [this tutorial](https://www.code-fabrik.com/2018/09/25/traefik-ein-reverse-proxy-f%C3%BCr-docker/). 

The main idea is:
- We require users of the registry to have a traefik instance running inside a docker container on the docker host (installation may be part of the setup script)
- When deploying new containers (WebApi, WebApp, Crawler), we label the containers with the *unique* host URL as required by traefik (e.g. ```traefik.frontend.rule=Host:api1.delphi.cs.upb.de```). We can use the instance id to generate unique URLs. We store the generated URL in the ```host``` attribute of the respective instance object
- Traefik will automatically detect these containers and provide access to them. Since by default there is no DNS name resolution, a call to the ```/version``` endpoint of a container via traefik (running at 127.0.0.1:80) will look like this:

 ```
      curl 127.0.0.1:80/version -H "Host: api1.delphi.cs.upb.de"
```

This way the management application will be able to directly call instances based on their ```host``` attribute and the Docker host URL. If instances need to be reachable via their host name, the respective DNS settings need to be configured accordingly (e.g. for public webapi / webapp).

I will try to implement a prototype of this for the upcoming [milestone](https://github.com/delphi-hub/delphi-registry/milestone/2).
